src/lib/requestmanager.ts

```diff
@@ -72,23 +95,42 @@ export class RequestManager {
 			this.#clearTimeout = null
 		}
 		const now = Date.now()
-		const qty = this.#entries.size
-		this.#entries.forEach(({ deferred, createdAt }, id) => {
-			if (force || now - createdAt > this.#timeout) {
-				deferred.reject(new Error('ERR_TIMEOUT'))
-				this.#entries.delete(id)
+		const qty = this.#requestsByTime.length
+		// filter() is usually faster than splice() for small-ish arrays
+		this.#requestsByTime = this.#requestsByTime.filter((request) => {
+			if (!this.#requestsById.has(request.invokeId)) {
+				// Request has already been resolved or expired
+				return false
+			}
+			if (force || request.expiresAt <= now) {
+				// Request has timed out or we forcefully time it out
+				request.deferred.reject(new Error('ERR_TIMEOUT'))
+				this.#requestsById.delete(request.invokeId)
+				return false
 			}
+			// Request is still pending
+			return true
 		})
-		debug(`Cleared ${qty - this.#entries.size} entries.`)
-		debug(`There are ${this.#entries.size} entries pending.`)
-		if (this.#entries.size > 0) {
+		assert(
+			this.#requestsById.size === this.#requestsByTime.length,
+			`Index size mismatch  ${this.#requestsById.size} !== ${this.#requestsByTime.length}`,
+		)
+		debug(`Cleared ${qty - this.#requestsByTime.length} entries.`)
+		debug(`There are ${this.#requestsByTime.length} entries pending.`)
+		if (!force) {
 			this.#scheduleClear()
 		}
 	}
 
 	#scheduleClear() {
-		if (this.#clearTimeout === null) {
-			this.#clearTimeout = setTimeout(this.clear, this.#timeout)
+		if (this.#clearTimeout === null && this.#requestsByTime.length > 0) {
+			// setTimeout() effectively becomes like setImmediate() if the delay value is 0 or negative
+			this.#clearTimeout = setTimeout(
+				this.clear,
+				// We add 100 milliseconds to ensure that we can never schedule more than
+				// 10 timeouts per second even in worst-case bursts scenarios
+				this.#requestsByTime[0].expiresAt - Date.now() + 100,
```

Comment:
I would firstly store this to a var, do some checks on them for example ensure it is grater then 0, print a log about the new timeout set and the set the timeout

Reply:
Done in https://github.com/bacnet-js/client/pull/52/commits/13f2b1dc9f0f5ea921e6c58c070a0f1e469a2630